Xamarin.Forms.Core/MediaElement.cs

```diff
@@ -0,0 +1,267 @@
+	public sealed class MediaElement : View
+	{
+		public static readonly BindableProperty AreTransportControlsEnabledProperty =
+		  BindableProperty.Create(nameof(AreTransportControlsEnabled), typeof(bool), typeof(MediaElement), false);
+
+		public static readonly BindableProperty AspectProperty =
+		  BindableProperty.Create(nameof(Aspect), typeof(Aspect), typeof(MediaElement), Aspect.AspectFit);
+
+		public static readonly BindableProperty AutoPlayProperty =
+		  BindableProperty.Create(nameof(AutoPlay), typeof(bool), typeof(MediaElement), true);
+
+		public static readonly BindableProperty BufferingProgressProperty =
+		  BindableProperty.Create(nameof(BufferingProgress), typeof(double), typeof(MediaElement), 0.0);
+
+		public static readonly BindableProperty IsLoopingProperty =
+		  BindableProperty.Create(nameof(IsLooping), typeof(bool), typeof(MediaElement), false);
+
+		public static readonly BindableProperty KeepScreenOnProperty =
+		  BindableProperty.Create(nameof(KeepScreenOn), typeof(bool), typeof(MediaElement), false);
+
+		public static readonly BindableProperty SourceProperty =
+		  BindableProperty.Create(nameof(Source), typeof(Uri), typeof(MediaElement));
+
+		public static readonly BindableProperty CurrentStateProperty =
+		  BindableProperty.Create(nameof(CurrentState), typeof(MediaElementState), typeof(MediaElement), MediaElementState.Closed);
+
+		public static readonly BindableProperty PositionProperty =
+		  BindableProperty.Create(nameof(Position), typeof(TimeSpan), typeof(MediaElement), TimeSpan.Zero, validateValue: ValidatePosition);
+
+		static bool ValidatePosition(BindableObject bindable, object value)
+		{
+			MediaElement element = bindable as MediaElement;
+			if (element != null)
+			{
+				if (element._renderer != null)
+				{
+					element._renderer.Seek((TimeSpan)value);
+				}
+			}
+
+			return true;
+		}
+
+		
+		IMediaElementRenderer _renderer = null;
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SetRenderer(IMediaElementRenderer renderer)
+		{
+			_renderer = renderer;
+		}
+		
+		public bool AreTransportControlsEnabled
+		{
+			get { return (bool)GetValue(AreTransportControlsEnabledProperty); }
+			set { SetValue(AreTransportControlsEnabledProperty, value); }
+		}
+		
+		public bool AutoPlay
+		{
+			get { return (bool)GetValue(AutoPlayProperty); }
+			set { SetValue(AutoPlayProperty, value); }
+		}
+
+		public double BufferingProgress
+		{
+			get { return (double)GetValue(BufferingProgressProperty); }
+		}
+		
+		public bool CanSeek
+		{
+			get { return Source != null && NaturalDuration.HasValue; }
+		}
+		
+		public bool IsLooping
+		{
+			get { return (bool)GetValue(IsLoopingProperty); }
+			set { SetValue(IsLoopingProperty, value); }
+		}
+		
+		public bool KeepScreenOn
+		{
+			get { return (bool)GetValue(KeepScreenOnProperty); }
+			set { SetValue(KeepScreenOnProperty, value); }
+		}
+
+		public TimeSpan? NaturalDuration
+		{
+			get
+			{
+				if (_renderer != null)
```

Comment:
Core should not know about the renderer

Reply:
I'm completely rewriting this to remove the IMediaElementRenderer interface and all the renderer references from the MediaElement. I've implemented instead an IMediaElementController on the MediaElement which the renderer can use to set things in the control without exposing them to the developer just consuming the MediaElement. I've used the WebView as a guide for the architecture.